### PR TITLE
Reuse reset token

### PR DESCRIPTION
**Pull request by rma4ok**
### Summary
This PR adds 2 new features allowing precise configuration over the password reset functionality.

`options.resetAttempts = 10` will not immediately delete the existing token from DB if the user submits an invalid token
`options.reuseResetToken = true` allows to save reset token to DB without hashing it and use it again when the user resets the password many times within a short period of time

These are absolutely non breaking changes and default values for the options are providing the same behavior as it is right now. But I would suggest to use `resetAttempts: 10` instead of `0` as it doesn't impose any security concerns and enhances user experience overall

- [x] Tell us about the problem your pull request is solving.
Actual users having problems with resetting their password. Different scenarios are described below.
- [x] Are there any open issues that are related to this?
No
- [x] Is this PR dependent on PRs in other repos?
No

If so, please mention them to keep the conversations linked together.

### Real-world scenarios
Current behavior:
An email delivery service might sometimes take 3-5 minutes to deliver the password reset email. After waiting for 1-2 minutes the user might want to try to reset the password again. The existing implementation will generate a new token and replace the previous token in DB. A minute later the email from the first reset is arriving and the user trying to reset the password with the token that is not valid anymore (got replaced by the 2nd token). The invalid token will throw an error that removes 2nd reset token from DB. Another minute later the second email arrives, but its token is also not valid anymore (got deleted because of 1st error). The user needs to reset the password again and the story might repeat.

Desired behavior:
When the user resets the password twice both emails would work as expected (achieved with `reuseResetToken`)
When the user clicks on some old email and tries to reset it will not clear the current token (achieved with `resetAttempts`)

 :heart:

### CODE REVIEW

**Comment by rma4ok** [comment]
Hi @claustres! Thank you for the valuable feedback!
Regarding `testMode: true` I think I've picked it from the test I've followed as an example. After investigation, I didn't find this option to do anything, so I've removed it from all places.

I also had in mind the security concern you are describing and that why `removeTokenOnError` was introduces as an option, so it would be a developer's responsibility to use this option and decrease `resetDelay` or use API request limiter.

But after thinking further I've realized that this issue can be cleanly resolved within this module and replaced `removeTokenOnError: true` with `resetAttempts: 10`. So we can configure the amount of failed attempts before the current token is deleted from the user object in the database.

**Comment by claustres** [comment]
Good catch Kiryl, a pleasure to collaborate as a reviewer, sometimes a simple remark makes you process a new idea! Your approach seems pretty elegant to me, and no breaking changes by default, which is great for end-users.

Waiting for CI but looks good to merge.

**Comment by rma4ok** [comment]
@claustres before we merge, do you think `resetAttempts` should be `0` or `10` by default?
`0` would produce behaviour identical to the current one, `10` will enhance the user experience

**Comment by claustres** [comment]
Always try to keep backward compatibility IMHO, above all with security concerns, we should not introduce a potential threat without letting users know, so `0`.

Before merging I also ask about any feedback from current maintainers: @OnnoGabriel @TheSinding @dallinbjohnson @bjohnson-ionrev  ?

**Comment by TheSinding** [comment]
I'll look at the pullrequest when i get home tonight !

File: src/send-reset-pwd.js
Comment (TheSinding) [inline src/send-reset-pwd.js:33]: This isn't very readable.
Could you make it more readable ? 

And why do you need to check if the `user1.resetToken.includes('___')`?

**Review by TheSinding — CHANGES_REQUESTED** [review]
Nice work !

One question:
Why would we need to reuse a token within a given timeframe, and not just generate a new one ?

File: src/send-reset-pwd.js
Comment (rma4ok) [reply src/send-reset-pwd.js]: Hey @TheSinding ! I've changed the lines and made the comment more explicit. 
`includes('___')` to make sure that we have a plain token in the user object, not the hashed one. You can see the same check [here](https://github.com/feathersjs-ecosystem/feathers-authentication-management/blob/master/src/helpers/deconstruct-id.js#L7). Some tests are failing without it. And in the real-world scenario, if someone  changes the `reuseResetToken` setting in production while some user already requested the token.

File: src/reset-password.js
Comment (rma4ok) [reply src/reset-password.js]: Changed as you requested

**Comment by rma4ok** [comment]
Thank you for reviewing @TheSinding!
> Why would we need to reuse a token within a given timeframe, and not just generate a new one?

Generating a new token every time causes problems in real-world scenarios. It takes time for an email or SMS to arrive and a newly generated token deletes the previous one.

Example:
A user requests pass reset 1st time, then waits for 1 min and requests pass reset 2nd time before 1st email arrives. Then 1st email arrives, the user clicks on it and sees the error (the token is overridden by 2nd reset request). User requests 3rd pass reset, then 2nd email arrives which is now invalid too. Then the user contacts the support team and says pass reset is not working.

Please read the full scenario in the PR description. Also as pass reset emails are similar, GMAIL will show only the first email body and collapse consequent emails into `...` button. So users might just keep clicking on the 1st email.

File: src/reset-password.js
Comment (TheSinding) [reply src/reset-password.js]: That makes sense :) 
Thanks mate :)

File: src/send-reset-pwd.js
Comment (TheSinding) [reply src/send-reset-pwd.js]: That makes sense, thanks for the clarification mate! :+1:

**Comment by TheSinding** [comment]
Sorry, that is great. 
I hadn't thought about that scenario. 
That's great, good PR mate ! :tada:

**Review by TheSinding — APPROVED** [review]

**Comment by rma4ok** [comment]
@claustres Thank you for merging this in! 
Do you have some idea when we can expect a new npm release, so we can integrate the official package with our service?
Cheers ✌

**Comment by claustres** [comment]
Will probably have a look next week.

**Pull request merged**
